#!/usr/bin/env Rscript
# Written by Noah Winters, Aug 2021
# This script identifies NLR clusters on each chromosome based on a window size cutoff (in genes)
# The orthogroup file is generated by PlantTribes
# The NLR density file is created by generate_ideogram_files.sh 
# All it takes as input is:
# (1) Genotype ID (as it appears in file), 
# (2) Tab-delimited orthogroup file like that generated by PlantTribes. Gene in one column, orthogroup assignment in the other
# (3) Tab-delimited list of NLR genes and their positions for the NLR-dense regions, tab delimited -- includes Chr, Window_Start, Window_Stop, Gene_Start, Gene_Stop, Gene

suppressPackageStartupMessages(library("argparse"))
suppressPackageStartupMessages(library("tidyverse"))
suppressPackageStartupMessages(library("valr"))


# create parser object
parser <- ArgumentParser()

# specify our desired options 
# by default ArgumentParser will add an help option 
parser$add_argument("-g", "--genotype", type="character", default="None", 
                    help="Genotype of interest [default %(default)s]")
parser$add_argument("--orthogroup_file", type = "character", default="None", 
                    help = "Tab-delimited orthogroup file like that generated by PlantTribes. Gene in one column, orthogroup assignment in the other. [default %(default)s]")
parser$add_argument("--nlr_density_file", type = "character", default="None",
                    help="Tab-delimited file containing NLR hotspots across the genome, like that generated by generate_ideogram_files.sh, ending in {genotype}_nlr_denseRegions_genes.txt. [default %(default)s]")
parser$add_argument("--cluster_size", type = "integer", default=3,
                    help="Minimum size for NLR clusters (in # of genes) [default %(default)s]")


# Get command line options, if help option encountered print help and exit,
# Otherwise if options not found on command line then set defaults, 
args <- parser$parse_args()

# Input Arguments 
gntype = args$genotype
ortho_file <- args$orthogroup_file
nlr_density <- args$nlr_density_file
N <- args$cluster_size

# Make dplyr less noisy 
options(dplyr.summarise.inform = FALSE)

# Table of orthogroup classifications for each NLR
orthogroups <- read.table(ortho_file, header = TRUE, stringsAsFactors = FALSE, sep = '\t') %>%
  setNames(., nm = c("Gene", "Orthogroup"))

# Parse to handle special cases, i.e. Criollo and Matina
if (gntype == "Criollo"){
  orthogroups <-
    orthogroups %>%
    .[grep("^Tc.*", .$Gene),] %>%
    mutate(Gene = gsub("\\.[0-9].*", "", Gene)) %>%
    mutate(Gene = gsub("_p", "_g", Gene))
} else if (gntype == "Matina"){
  orthogroups <-
    orthogroups %>%
    .[grep("^Thecc.*", .$Gene),] %>%
    mutate(Gene = gsub("\\.1\\.p", "", Gene))
} else {
  orthogroups <-
    orthogroups %>%
    .[grep(gntype, .$Gene),] %>%
    mutate(Gene = gsub("\\.[0-9].*", "", Gene))
}

# What are the highly dense NLR regions?
denseRegions <- read.table(nlr_density, header = TRUE, stringsAsFactors = FALSE, sep = '\t') %>%
  inner_join(orthogroups, by = "Gene")


##########################
### Cluster Membership ###
##########################
# Merge windows if they overlap, i.e. adjacent regions with NLRs will be merged
# Merging will continue until there are no more adjacent windows
# These represent NLR clusters 
nlr_clusters <- 
  denseRegions %>% 
  dplyr::select(c("chrom" = Chr, 
                  "start" = Window_Start, 
                  "end" = Window_Stop)) %>% 
  valr::bed_merge(.) %>% 
  group_by(chrom) %>%
  mutate(id = paste0(chrom, "_", "cluster_", 1:n()))

# Merge NLR clusters with genes in each cluster, count how many there are 
# Only include clusters of size N or greater
denseRegions_clusters <- 
  denseRegions %>% 
  dplyr::select(c("chrom" = Chr, 
                  "start" = Gene_Start, 
                  "end" = Gene_Stop, 
                  "id" = Gene)) %>%
  valr::bed_intersect(., nlr_clusters, suffix = c("_gene","_cluster")) %>% 
  inner_join(orthogroups, by = c("id_gene" = "Gene")) %>% 
  group_by(start_cluster, end_cluster) %>% 
  add_tally(name = "num_genes_cluster") %>% 
  filter(num_genes_cluster >= N)

# Write out clusters and the genes contained in each
denseRegions_clusters %>%
  write.csv(., paste0(gntype,"_nlr_cluster_membership.csv"),
            quote = FALSE,
            row.names = FALSE)


#########################
### Cluster Ideograms ###
#########################
# Orthogroups belonging to each cluster, used for plotting ideograms
denseRegions_ortho <-
  denseRegions_clusters %>%
  dplyr::select(c(chrom, start_cluster, end_cluster, id_gene, Orthogroup)) %>%
  group_by(chrom, start_cluster, end_cluster) %>%
  dplyr::summarize(unique(Orthogroup), ) %>%
  tally(name = "Num_Unique_Orthos") %>% 
  add_column(Type = ifelse(.$Num_Unique_Orthos > 1, "Heterogeneous", "Homogeneous"),
             Shape = ifelse(Type == "Heterogeneous", "circle", "triangle")) %>%
  dplyr::select(c(Type, Shape, chrom, start_cluster, end_cluster)) %>%
  setNames(nm = c("Type", "Shape", "Chr", "Start", "End")) %>%
  distinct() %>%
  write.csv(., paste0(gntype,"_cluster_pos_perChr_ideogram.csv"),
          quote = FALSE,
          row.names = FALSE)


######################
### Cluster Totals ###
######################
# Write out file containing # NLR clusters per chromosome
denseRegions_clusters %>%
  dplyr::select(c(chrom, start_cluster, end_cluster,num_genes_cluster)) %>%
  distinct() %>%
  group_by(chrom) %>%
  dplyr::summarize(
    Mean_Clust_Size = round(mean(num_genes_cluster), 2),
    Num_Clusters = n(), 
  ) %>%
  write.csv(., paste0(gntype,"_cluster_totals_perChr.csv"),
          quote = FALSE,
          row.names = FALSE)

cat(paste0("Finished processing: ", gntype))
cat("\n")
